rfcs/text/0021-threat-enrichment.md

```diff
@@ -39,12 +50,130 @@ Stage 2: Add or update all remaining field definitions. The list should now be e
 Stage 1: Describe at a high-level how these field changes will be used in practice. Real world examples are encouraged. The goal here is to understand how people would leverage these fields to gain insights or solve problems. ~1-3 paragraphs.
 -->
 
+### Adding threat intelligence match/enrichment to another document
+
+If it is determined that an event matches a given indicator, that event can be enriched with said indicator. Presently, Indicator Match Detection rules will perform that enrichment automatically, but ad hoc/manual enrichment is an expected feature that analysts will leverage in the future.
+
+#### Example document
+
+```json5
+{
+  "process": {
+    "name": "svchost.exe",
+    "pid": 1644,
+    "entity_id": "MDgyOWFiYTYtMzRkYi1kZTM2LTFkNDItMzBlYWM3NDVlOTgwLTE2NDQtMTMyNDk3MTA2OTcuNDc1OTExNTAw",
+    "executable": "C:\\Windows\\System32\\svchost.exe"
+  },
+  "message": "Endpoint file event",
+  "@timestamp": "2020-11-17T19:07:46.0956672Z",
+  "file": {
+    "path": "C:\\Windows\\Prefetch\\SVCHOST.EXE-AE7DB802.pf",
+    "extension": "pf",
+    "name": "SVCHOST.EXE-AE7DB802.pf",
+    "hash": {
+      "sha256": "0c415dd718e3b3728707d579cf8214f54c2942e964975a5f925e0b82fea644b4"
+    }
+  },
+  "threat": {
+    "enrichments": [
+      {
+        // Each enrichment is added as a nested object under `threat.enrichments.*`
+        // Copy all the object indicators under `indicator.*`, providing full context
+        "indicator": {
+          "marking": {
+            "tlp": "WHITE"
+          },
+          "first_seen": "2020-10-01",
+          "last_seen": "2020-11-01",
+          "sightings": 4,
+          "type": ["sha256", "md5", "file_name", "file_size"],
+          "description": "file last associated with delivering Angler EK"
+        },
+        // event and file fields are copied from the indicator doc, if relevant
+        "event": {
+          "provider": "Abuse.ch",
+          "dataset": "threatintel.abusemalware",
+          "module": "threatintel"
+        },
+        "file": {
+          "hash": {
+            "sha256": "0c415dd718e3b3728707d579cf8214f54c2942e964975a5f925e0b82fea644b4",
+            "md5": "1eee2bf3f56d8abed72da2bc523e7431"
+          },
+          "size": 656896,
+          "name": "invoice.doc"
+        },
+        /* `matched` will provide context about which of the indicators above matched on this
+              particular enrichment. If multiple matches for this indicator object, this could
+              be a list */
+        "matched": {
+          "atomic": "0c415dd718e3b3728707d579cf8214f54c2942e964975a5f925e0b82fea644b4",
```

Comment:
I'm not convinced that `matched.atomic` and `matched.field` are sufficient to cover more sophisticated matching mechanisms than an exact match: if e.g. a user wanted to write an indicator match rule with a partial match, or if there were a more sophisticated indicator that itself represented a wildcard/regex value, then the value of the indicator would not be identical to the value from the event and so we may want two fields here.

Similarly, we would not be able to reproduce the exact conditions of the match with only one `field` value. While I've been told that the indicator field being matched upon should be self-evident, it seems safer to explicitly state it in another field, and I wanted to bring it up one more time before I shut up about it 😉 .

In general, I'm viewing these `matched.*` fields as the answer to both HOW and WHY a given event was enriched, so keep that in mind and/or correct me on that thinking.

Reply:
> a user wanted to write an indicator match rule with a partial match, or if there were a more sophisticated indicator that itself represented a wildcard/regex value, then the value of the indicator would not be identical to the value from the event

How are you envisioning this? Are you thinking that there is a mechanical way needed to match _like_ fields. I can think of a use-case, DGAs. 

So, if you know an aggressor uses a specific domain structure for C2 (like `abc123\.12345abcdef\.xyz`), you'd want to be able to match if they use
```
abc123\.12345abcdef\.xyz
plo958\.59874qwersd\.xyz
lje456\.01258iekduh\.xyz
```
So, you'd need a way to match `url.full:/[aA-zZ]{3}[0-9]{3}\.[0-9]{5}[aA-zZ]{6}\.xyz MATCHES threat.indicator.url.full`?

Or am I off?

Reply:
@peasead right, the idea is that with anything other than an exact match, the values for the LHS and RHS of the match are going to be different, and we don't have fields to account for both of those right now.

IPs is another example, where one could specify a CIDR block instead of a single IP address.

What is SOP for DGAs within threat intel, currently? Is there any attempt to generalize the pattern/algorithm generating those values?

Reply:
I'm not saying it's a must now.

There are some [experimental machine learning jobs](https://github.com/elastic/detection-rules/releases/tag/ML-experimental-detections-20200506-3) for DGAs.

Reply:
I think we can continue with the implicit "exact value match" semantics for this RFC. If in the future we need to support the aforementioned functionality, I think that we can do so with the addition of two new fields as discussed.